src/blas/backends/cublas/cublas_scope_handle.cpp

```diff
@@ -0,0 +1,145 @@
+    CUcontext primary;
+    cuDevicePrimaryCtxRetain(&primary, cudaDevice);
+    bool isPrimary = primary == desired;
+    cuDevicePrimaryCtxRelease(cudaDevice);
+    if (original_ != desired) {
+        // Sets the desired context as the active one for the thread
+        CUDA_ERROR_FUNC(cuCtxSetCurrent, err, desired);
+        // No context is installed and the suggested context is primary
+        // This is the most common case. We can activate the context in the
+        // thread and leave it there until all the PI context referring to the
+        // same underlying CUDA primary context are destroyed. This emulates
+        // the behaviour of the CUDA runtime api, and avoids costly context
+        // switches. No action is required on this side of the if.
+        needToRecover_ = !(original_ == nullptr && isPrimary);
+    }
+}
+
+CublasScopedContextHandler::~CublasScopedContextHandler() noexcept(false) {
+    if (needToRecover_) {
+        CUresult err;
+        CUDA_ERROR_FUNC(cuCtxSetCurrent, err, original_);
+    }
+}
+
+void ContextCallback(void *userData) {
+    auto *ptr = static_cast<std::atomic<cublasHandle_t> **>(userData);
+    if (!ptr) {
+        return;
+    }
+    if (*ptr != nullptr) {
+        auto handle = (*ptr)->exchange(nullptr);
+        if (handle != nullptr) {
+            cublasStatus_t err1;
+            CUBLAS_ERROR_FUNC(cublasDestroy, err1, handle);
+            handle = nullptr;
+        }
+        delete *ptr;
+        *ptr = nullptr;
+    }
+}
+
+cublasHandle_t CublasScopedContextHandler::get_handle(const cl::sycl::queue &queue) {
+    auto piPlacedContext_ = reinterpret_cast<pi_context>(placedContext_.get());
+    CUstream streamId     = get_stream(queue);
+    cublasStatus_t err;
+    auto it = handle_helper.cublas_handle_mapper_.find(piPlacedContext_);
+    if (it != handle_helper.cublas_handle_mapper_.end()) {
+        if (it->second == nullptr) {
+            handle_helper.cublas_handle_mapper_.erase(it);
+        }
+        else {
+            auto handle = it->second->load();
+            if (handle != nullptr) {
+                cudaStream_t currentStreamId;
+                CUBLAS_ERROR_FUNC(cublasGetStream, err, handle, &currentStreamId);
+                if (currentStreamId != streamId) {
+                    CUBLAS_ERROR_FUNC(cublasSetStream, err, handle, streamId);
+                }
+                return handle;
+            }
+            else {
+                handle_helper.cublas_handle_mapper_.erase(it);
+            }
+        }
+    }
+
+    cublasHandle_t handle;
+
+    CUBLAS_ERROR_FUNC(cublasCreate, err, &handle);
+    CUBLAS_ERROR_FUNC(cublasSetStream, err, handle, streamId);
+
+    auto insert_iter = handle_helper.cublas_handle_mapper_.insert(
+        std::make_pair(piPlacedContext_, new std::atomic<cublasHandle_t>(handle)));
+
+    auto ptr = &(insert_iter.first->second);
+
+    sycl::detail::pi::contextSetExtendedDeleter(placedContext_, ContextCallback, ptr);
```

Comment:
For some reasons I don't see this symbol in libsycl.so (built from https://github.com/intel/llvm/commit/2d71c8ea04b9b087f0c5ede0745189118a746a5f), the only one I found is not visible at link time:
```
00000000002148b0 t cl::sycl::detail::pi::contextSetExtendedDeleter(cl::sycl::context const&, void (*)(void*), void*)
```

Reply:
LLVM compiler was built with following flags " --cuda --shared-libs -t release"
Ubuntu 18.04 with CUDA 10.2
Please let me know if I missed something

Reply:
Hi @mkrainiuk, Thanks for reporting the issue. `__SYCL_EXPORT` attribute was missing from contextSetExtendedDeleter definition, preventing it from being publically visible. [PR #1531](https://github.com/intel/llvm/pull/1531) will fix it.

Reply:
Hi @mehdi-goli, I've checked the fix on my side, it works now. Thank you!